server/src/main/java/org/elasticsearch/health/node/action/TransportHealthNodeAction.java

```diff
@@ -0,0 +1,141 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.health.node.selection.HealthNode;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskCancelledException;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportService;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * A base class for operations that need to be performed on the health node.
+ */
+public abstract class TransportHealthNodeAction<Request extends ActionRequest, Response extends ActionResponse> extends
+    HandledTransportAction<Request, Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportHealthNodeAction.class);
+
+    protected final TransportService transportService;
+    protected final ClusterService clusterService;
+    protected final ThreadPool threadPool;
+    protected final String executor;
+
+    private final Writeable.Reader<Response> responseReader;
+
+    protected TransportHealthNodeAction(
+        String actionName,
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        Writeable.Reader<Request> request,
+        Writeable.Reader<Response> response,
+        String executor
+    ) {
+        super(actionName, true, transportService, actionFilters, request);
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.executor = executor;
+        this.responseReader = response;
+    }
+
+    protected abstract void healthOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener)
+        throws Exception;
+
+    @Override
+    protected void doExecute(Task task, final Request request, ActionListener<Response> listener) {
+        ClusterState state = clusterService.state();
+        logger.trace("starting to process request [{}] with cluster state version [{}]", request, state.version());
+        if (task != null) {
+            request.setParentTask(clusterService.localNode().getId(), task.getId());
+        }
+        new AsyncSingleAction(task, request, listener).doStart(state);
+    }
+
+    class AsyncSingleAction {
```

Comment:
Do we need a class here? The logic it provides could be a method?

Maybe I'm missing something but seems to just wrap the arguments `doExecute` has?

Reply:
Cleaned up.